typescript/sdk/src/token/read.ts

```diff
@@ -0,0 +1,138 @@
+    protected readonly chain: ChainName,
+    protected readonly concurrency: number = DEFAULT_CONTRACT_READ_CONCURRENCY,
+  ) {
+    this.provider = this.multiProvider.getProvider(chain);
+    this.evmHookReader = new EvmHookReader(multiProvider, chain, concurrency);
+    this.evmIsmReader = new EvmIsmReader(multiProvider, chain, concurrency);
+  }
+
+  /**
+   * Derives the configuration for a Hyperlane ERC20 router contract at the given address.
+   *
+   * @param address - The address of the Hyperlane ERC20 router contract.
+   * @returns The configuration for the Hyperlane ERC20 router.
+   *
+   */
+  async deriveWarpRouteConfig(
+    address: Address,
+  ): Promise<Omit<ERC20RouterConfig, 'type' | 'gas'>> {
+    const fetchedBaseMetadata = await this.fetchBaseMetadata(address);
+    const fetchedTokenMetadata = await this.fetchTokenMetadata(
+      fetchedBaseMetadata.wrappedTokenAddress,
+    );
+
+    const results: Omit<ERC20RouterConfig, 'type' | 'gas'> = {
+      ...fetchedBaseMetadata,
+      ...fetchedTokenMetadata,
+    };
+
+    if (
+      fetchedBaseMetadata.interchainSecurityModule !==
+      ethers.constants.AddressZero
+    ) {
+      results.interchainSecurityModule =
+        await this.evmIsmReader.deriveIsmConfig(
+          fetchedBaseMetadata.interchainSecurityModule,
+        );
+    }
+    // @todo add after https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3667 is fixed
+    // if (fetchedBaseMetadata.hook !== ethers.constants.AddressZero) {
+    //   results.hook = await this.evmHookReader.deriveHookConfig(
+    //     fetchedBaseMetadata.hook,
+    //   );
+    // }
+
+    return results;
+  }
+
+  /**
+   * Fetches the base metadata for a Warp Route contract.
+   *
+   * @param routerAddress - The address of the Warp Route contract.
+   * @returns The base metadata for the Warp Route contract, including the mailbox, owner, wrapped token address, hook, and interchain security module.
+   */
+  async fetchBaseMetadata(
+    routerAddress: Address,
+  ): Promise<WarpRouteBaseMetadata> {
+    const warpRoute = HypERC20Collateral__factory.connect(
+      routerAddress,
+      this.provider,
+    );
+    const [
+      mailbox,
+      owner,
+      wrappedTokenAddress,
+      hook,
+      interchainSecurityModule,
+    ] = await Promise.all([
+      warpRoute.mailbox(),
+      warpRoute.owner(),
+      warpRoute.wrappedToken(),
+      warpRoute.hook(),
+      warpRoute.interchainSecurityModule(),
+    ]);
+
+    return {
+      mailbox,
+      owner,
+      wrappedTokenAddress,
+      hook,
+      interchainSecurityModule,
+    };
+  }
+
+  /**
+   * Fetches the metadata for a token address.
+   *
+   * @param tokenAddress - The address of the token.
+   * @returns A partial ERC20 metadata object containing the token name, symbol, total supply, and decimals.
+   */
+  async fetchTokenMetadata(
+    tokenAddress: Address,
+  ): Promise<Partial<ERC20Metadata>> {
```

Comment:
it doesn't look like you're actually returning a Partial because the only missing field `scale` is optional anyway

Reply:
Fixed